base_layer/core/src/base_node/states/block_sync.rs

```diff
@@ -67,213 +96,309 @@ impl BlockSyncInfo {
+        } else {
+            trace!(
+                target: LOG_TARGET,
+                "Block hash {} is common between our chain and the network.",
+                local_block_hash.to_hex()
+            );
+        }
+
+        info!(target: LOG_TARGET, "Synchronize missing blocks.");
+        for height in sync_height..=network_tip_height {
+            request_and_add_block(db, comms, config, sync_peers, height).await?;
+        }
+        return Ok(());
+    }
+    Err(BlockSyncError::EmptyBlockchain)
+}
+
+// Perform a basic check to determine if a chain split has occurred between the local and network chain.
+async fn check_chain_split(
+    comms: &mut OutboundNodeCommsInterface,
+    config: &BlockSyncConfig,
+    sync_peers: &[NodeId],
+    local_tip_height: u64,
+    network_tip_height: u64,
+    local_block_hash: &BlockHash,
+) -> Result<bool, BlockSyncError>
+{
+    Ok(if network_tip_height >= local_tip_height {
+        let header = request_header(comms, config, sync_peers, local_tip_height).await?;
+        *local_block_hash != header.hash()
+    } else {
+        true
+    })
+}
+
+// Find the block height where the chain split occurs. The chain split height is the height of the first block that is
+// not common between the local and network chains.
+async fn find_chain_split_height<B: BlockchainBackend + 'static>(
+    db: &BlockchainDatabase<B>,
+    comms: &mut OutboundNodeCommsInterface,
+    config: &BlockSyncConfig,
+    sync_peers: &[NodeId],
+    tip_height: u64,
+) -> Result<u64, BlockSyncError>
+{
+    for block_nums in (1..=tip_height)
+        .rev()
+        .collect::<Vec<u64>>()
+        .chunks(config.header_request_size)
+    {
+        let headers = request_headers(comms, config, sync_peers, block_nums).await?;
+        for header in headers {
+            // Check if header is linked to local chain
+            if let Ok(prev_header) = async_db::fetch_header_with_block_hash(db.clone(), header.prev_hash.clone()).await
             {
-                trace!(
+                if prev_header.height + 1 == header.height {
+                    return Ok(header.height);
+                } else {
+                    return Err(BlockSyncError::InvalidChainLink);
+                }
+            }
+        }
+    }
+    Err(BlockSyncError::ForkChainNotLinked)
+}
+
+// Request a block from a remote sync peer and attempt to add it to the local blockchain.
+async fn request_and_add_block<B: BlockchainBackend + 'static>(
+    db: &BlockchainDatabase<B>,
+    comms: &mut OutboundNodeCommsInterface,
+    config: &BlockSyncConfig,
+    sync_peers: &[NodeId],
+    height: u64,
+) -> Result<(), BlockSyncError>
+{
+    for attempt in 0..config.max_add_block_retry_attempts {
+        let block = request_block(comms, config, sync_peers, height).await?;
+        let block_hash = block.hash();
+        match db.add_block(block.clone()) {
+            Ok(_) => {
+                debug!(
```

Comment:
This could be info, and also debug! printing the entire block out

Reply:
Change applied.